daffodil-core/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/StringLengthParsers.scala

```diff
@@ -86,8 +87,25 @@ trait StringOfSpecifiedLengthMixin extends PaddingRuntimeMixin with CaptureParsi
 
   protected final def parseString(start: PState): String = {
     val dis = start.dataInputStream
-    val maxLen = start.tunable.maximumSimpleElementSizeInCharacters
     val startBitPos0b = dis.bitPos0b
+    val bitLimit0b = dis.bitLimit0b
+
+    // We want to limit the maximum length passed into getSomeString since that function can
+    // pre-allocate a buffer that size even if it won't find that many characters. So we
+    // calculate the maximum number of characters that we could possibly decode based on the
+    // number of a available bits and the character set. Note that some character sets have
+    // variable width encodings, so we can't always know for sure how many characters will be
+    // found. But we can calculate a maximum based on the smallest possible code unit for the
```

Comment:
Feels backward to me. Don't you need the maximum width so you can allocate larger when there are variable-width characters possible and the string may contain N of those? I.e., so for UTF8 you want to allow the worst-case of 4-bytes per character, not 1?